Testing the single exercise checkout. Ready for request pull.

verbusharas opened the pull request:
Labas Pavel,

Bandau paskaitos metu aptartą darbų atsiskaitymo eigą. Kol kas Elvinas nepatvirtino kolaboracijos - tad kreipiuosi į Tave.

Prašau peržiūrėti mano Failų užduoties sprendimą. Čia sprendimas išskaidant į atskiras klases.
Šį uždavinį (ir dar kelis sekančius, kiek buvome spėję praeiti) atlikau seniau, kai dar dėstė Mantas.
Likusius uždavinius sukelsiu, kai prieisim atitinkamas paskaitas.

Ačiū.

verbusharas committed:
Testing the single exercise checkout. Ready for request pull.

pavelvrublevskij commented on Ex01_Files01/src/lt/verbus/ExportService.java:
```diff
@@ -0,0 +1,49 @@
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.*;
```
Eliminuojam asterisk'a, geriau rodom visas klases.

pavelvrublevskij commented on Ex01_Files01/src/lt/verbus/ExportService.java:
```diff
@@ -0,0 +1,49 @@
+            e.printStackTrace();
+        }
+
+        System.out.println("List of people sorted by -most received money- exported successfully to file: " + pathToFile);
```
Nors gali palikti ir cia, bet logiskiausiau butu iskelti i try bloka. Nes jei reikes refaktorinti ir iskelineti i atskira metoda tai nereikes gilintis ar kazka praleidai.

pavelvrublevskij commented on Ex01_Files01/src/lt/verbus/ExportService.java:
```diff
@@ -0,0 +1,49 @@
+
+        System.out.println("List of people sorted by -most received money- exported successfully to file: " + pathToFile);
+    }
+    public void exportSenders(Map<Integer, Person> people, String pathToFile){
```
Atskiram metodu blokus :)

pavelvrublevskij replied:
Metodas labai panasus i exportReceivers metoda. Gali pagalvoti kaip isvengti kodo dublikavimo.

pavelvrublevskij commented on Ex01_Files01/src/lt/verbus/ImportService.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    public Map<Integer, Payment> importPayments(String pathToFile) {
+        Map<Integer, Payment> transactions = new HashMap<>();
+        try (BufferedReader paymentReader = Files.newBufferedReader(Paths.get(pathToFile))) {
+            String line;
+            paymentReader.readLine(); //prašokama pirma eilutė su stulpelių pavadinimais
+            while ((line = paymentReader.readLine()) != null) {
+                String[] argumentsInLine = line.split(", ");
+                transactions.put(Integer.valueOf(argumentsInLine[0]),
+                        new Payment(Integer.valueOf(argumentsInLine[0]), Double.valueOf(argumentsInLine[1]), Integer.valueOf(argumentsInLine[2]), Integer.valueOf(argumentsInLine[3].trim())));
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+        System.out.println("List of completed transactions imported successfully. No of entries: " + transactions.size());
+        return transactions;
```
Bandom panaikinti kodo kartojima su pries tai klases metodu. Siulau atskiras imlementacijas (skirtuma) tiesiog iskelti i atskirus metodus. Kodas iskart taps patrauklesnis ir suprantamas.

pavelvrublevskij commented on Ex01_Files01/src/lt/verbus/Payment.java:
```diff
@@ -0,0 +1,31 @@
+package lt.verbus;
+
+public class Payment {
```
Tavo klase yra nekintanti (immutable) tai siulau pagal nekintancios klases taisykles prideti reikiamose vietose final raktazodi.

pavelvrublevskij commented on Ex01_Files01/src/lt/verbus/Person.java:
```diff
@@ -0,0 +1,39 @@
+        this.sentMoney = 0;
+    }
+
+    public void transactionIn(double amount) {
```
set...., set... get..., is..., of... priduoda programuotojui aiskumo. pvz setTransactionIn(double amount)

pavelvrublevskij reviewed (commented):
Pagrinde laukiu dublikatu panaikinimo ir galima approvinti

verbusharas committed:
removed asterisk to specify which exact classes are imported.
exportSenders and exportReceivers methods merged into reusable exportPeople method.

verbusharas committed:
unsucessfully tried to avoid code duplication :((. Both methods seem similar, however they return different Maps, prints different messages and reading/(object creating) algorithms differ.
I think, in order to reduce duplication - the whole logic of instantiating this class should be altered.

However some simplification was made by extracting Person and Payment creation methods as separate.

verbusharas committed:
Payment bean class updated according immutable class convention: immutable fields were declared final

verbusharas committed:
Person bean class improved by declaring immutable fields final and refactoring setters to have "set" prefix.

verbusharas committed:
Enum class created for clarity when selecting sorting order.

verbusharas committed:
Main class renamed due to conflicts in project structure accross different modules.

verbusharas commented:
Pavelai, labai ačiū už pastabas.
Vienos, deja, nepavyko įgyvendinti - t.y. ImportService kodo dublikavimas išliko. Bandžiau visokius variantus - ir su Generics ir su wildcards ir su castinimu ir su metodų atskyrimu - bet norint išvengti šito dublikato, manau reikia iš esmės perdaryti logiką, kaip kreipiasi main į service ir kaip kuriami objektai bei map'ai. Galbūt reikėtų papildomai daryti "fabrikiukus" kurie priklausomai nuo to ką perskaito faile, kreipiasi į atitinkamą fabrikėlį: argumentsInLine.length < 4 ? createPerson() : createPayment().

Todėl galų gale - nusprendžiau, kad importą palieku koks yra, o exportą sėkmingai sutvarkiau. 
Jau pushinau atnaujintą kodą.
